src/entity_ref.rs

```diff
@@ -294,3 +349,79 @@ impl<'a, T: Component> ComponentRef<'a> for &'a mut T {
 pub trait ComponentRefShared<'a>: ComponentRef<'a> {}
 
 impl<'a, T: Component> ComponentRefShared<'a> for &'a T {}
+
+struct ComponentBorrow<'a> {
+    archetype: &'a Archetype,
+    /// State index for `component` in `archetype`
+    state: usize,
+}
+
+impl<'a> ComponentBorrow<'a> {
+    unsafe fn for_component<T: Component>(
```

Comment:
A brief comment describing why this is unsafe (`index` is unchecked) wouldn't hurt